src/Task.cc

```diff
@@ -3671,6 +3671,12 @@ void Task::dup_from(Task *other) {
       create_mapping(this, remote_this, km);
       LOG(debug) << "Copying mapping into " << tid;
       if (!(km.flags() & MAP_SHARED)) {
+#if defined(__x86_64__) && defined(__SANITIZE_ADDRESS__)
```

Comment:
I think possibly masks a real issue that maybe should be investigated? I haven't looked into it myself yet.

Reply:
Yes this is just a workaround for asan mapping a really huge area, which cannot be allocated by plain malloc.
Maybe rr has to reproduce this mapping and copy just the "used" memory. Have not looked any further into this.

Reply:
Today I tried to modify the copy_mem_mapping to not try to allocate the whole chunk in one, instead to copy by e.g. 256 MB blocks. Unfortunately this did just start to fill RAM.
I guess if we really try to duplicate such a 20 TB "MAP_NORESERVE" mapping, a way is needed to tell pages which got accessed once and therefore are "in use" apart from these pages, which never got touched, and therefore the kernel did never allocate a hardware page for it. (If I understand this situation right?)

Reply:
One thing you can do here is read `/proc/<pid>/pagemap` and check if the page frame number for each anonymous page is zero. If it is, the page is all zeroes.

Reply:
Thanks for the hint.
I found the page frame number (PFN), bits 0-54 were all 0.
By testing the bit 63 "page present" seems to better show if something is in use.
Unfortunately testing the whole 20 TB increases the test time from 3 seconds to nearly 400 seconds.
And all in all there seem just around 126 pages of it in use.
So I guess this is not the final solution.
88792e9 contains my attempt to use /proc/pid/pagemap.

Reply:
I don't know why it's so slow for you. Here's a test program that reads 20TB worth of `/proc/self/pagemap` and counts the number of entries present: https://gist.github.com/rocallahan/a456903d7bc1480f27d16794ac8437c0
This runs in 22s:
```
$ gcc -O2 -o ~/tmp/test ~/tmp/test.c
$ time ~/tmp/test
Pages present: 0

real	0m22.243s
user	0m5.363s
sys	0m16.750s
```

Reply:
That program processes 1M pages per read (so 4GB instead of 1GB per read), and uses `pread()` to halve the number of syscalls, but I don't think that would explain why my program is 20x faster than your code.

Reply:
Your code calls `page_size()`, i.e. `sysconf(_SC_PAGE_SIZE)`, twice per chunk (and twice per page present!) ... maybe that's really slow? Or maybe building rr non-optimized, the `std::vector` operations (`size()` and `operator[]`) per page, are really slow.

Reply:
Thanks for your sample code.
I tried to do some modifications:
* 4m55: before
* 5m18: Assigned page_size() to a local variable (tested just once, seems I did something in parallel)
* 2m10: replaced the vector by a c-array

Then I changed the loop to your version (I hope it is ok to use it?).
And then I noticed, that I commented the `&& km.size() >= 0x100000000` and therefore used
`copy_mem_mapping_just_used` for all `MAP_NORESERVE` mappings,
which might not use holes and therefore get copied slowly page by page.
If I just use it for  I got it down to 35 seconds.

Reply:
Did you use pread instead of lseek/read? You should.

> copy_mem_mapping_just_used for all MAP_NORESERVE mappings,
> which might not use holes and therefore get copied slowly page by page.

Seems like the right thing to do is to identify runs of consecutive pages that are present, and handle each run with a single copy_mem_mapping call. I think you should use copy_mem_mapping_just_used for all anonymous mappings (which you're not checking for currently!), not just MAP_NORESERVE, and not just for big mappings.

Reply:
If I remember correctly I first found that just the first lseek with read was needed, so moved lseek before the loop.
But I could not see an improvement in time. Similar for removing lseek and changing read to pread.
Then I ripped out my version and used your lines.

I tried to add this identification of consecutive pages and call it for all sizes of ANONYMOUS|NORESERVE mappings.